src/api/component.js

```diff
@@ -3,21 +3,83 @@ import {
   connected as $connected,
   created as $created,
   props as $props,
-  renderer as $renderer,
-  rendererDebounced as $rendererDebounced,
+  renderer as $render,
+  rendererDebounced as $renderDebounced,
   rendering as $rendering,
   updated as $updated
 } from '../util/symbols';
 import data from '../util/data';
+import dashCase from '../util/dash-case';
 import debounce from '../util/debounce';
 import getAllKeys from '../util/get-all-keys';
 import getOwnPropertyDescriptors from '../util/get-own-property-descriptors';
 import getSetProps from './props';
+import initProps from '../lifecycle/props-init';
+import prop from '../util/prop';
 import syncPropToAttr from '../util/sync-prop-to-attr';
 
 const { HTMLElement } = window;
 
-function callConstructor (elem) {
+function syncPropsToAttrs (elem) {
+  const props = elem.constructor.props;
+  Object.keys(props).forEach((propName) => {
+    const prop = props[propName];
+    syncPropToAttr(elem, prop, propName, true);
+  });
+}
+
+// Ensures that definitions passed as part of the constructor are functions
+// that return property definitions used on the element.
+function ensurePropertyFunctions (Ctor) {
+  const props = Ctor.props;
+
+  return getAllKeys(props).reduce((descriptors, descriptorName) => {
+    descriptors[descriptorName] = props[descriptorName];
+    if (typeof descriptors[descriptorName] !== 'function') {
+      descriptors[descriptorName] = initProps(descriptors[descriptorName]);
+    }
+    return descriptors;
+  }, {});
+}
+
+// Ensures the property definitions are transformed to objects that can be used
+// to create properties on the element.
+function ensurePropertyDefinitions (Ctor) {
+  const props = ensurePropertyFunctions(Ctor);
+  return getAllKeys(props).reduce((descriptors, descriptorName) => {
+    descriptors[descriptorName] = props[descriptorName](descriptorName);
+    return descriptors;
+  }, {});
+}
+
+function createInitProps (Ctor) {
+  const props = ensurePropertyDefinitions(Ctor);
+
+  return (elem) => {
+    if (!props) {
```

Comment:
When exactly `props` can be falsy? `ensurePropertyDefinitions` will always return at least an empty object (since it's an initial value), isn't it?